File: VodArchiver/VodArchiver/VideoJobs/GenericFileJob.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Xml;
using VodArchiver.VideoInfo;

namespace VodArchiver.VideoJobs {
	class GenericFileJob : IVideoJob {
		public GenericFileJob( string url, StatusUpdate.IStatusUpdate statusUpdater = null ) {
			JobStatus = VideoJobStatus.NotStarted;
			StatusUpdater = statusUpdater == null ? new StatusUpdate.NullStatusUpdate() : statusUpdater;
			VideoInfo = new GenericVideoInfo() { Service = StreamService.RawUrl, VideoId = url };
			Status = "...";
		}

		public GenericFileJob( XmlNode node ) : base( node ) { }

		public override bool IsWaitingForUserInput => false;

		public override XmlNode Serialize( XmlDocument document, XmlNode node ) {
			node.AppendAttribute( document, "_type", "GenericFileJob" );
			return base.Serialize( document, node );
		}

		private static string StringToFilename( string basename, string extension ) {
			return Util.MakeStringFileSystemSafeBaseName( basename ) + "." + extension;
		}

		static long Counter = 0;

		public string GetTempFoldername() {
			long c = Interlocked.Increment(ref Counter);
			return "rawurl_" + c;
		}

		public string GetTargetFilename() {
			// best guess...
			string extension = Util.MakeStringFileSystemSafeBaseName( VideoInfo.VideoId.Trim().Split( '/' ).LastOrDefault().Split( '?' ).FirstOrDefault() ).Split( '_' ).LastOrDefault();
			if ( String.IsNullOrWhiteSpace( extension ) ) {
				extension = "bin";
			}

			string username = VideoInfo.Username.Trim();
			string title = String.IsNullOrWhiteSpace( VideoInfo.VideoTitle ) ? "file" : VideoInfo.VideoTitle.Trim();
			string datetime = VideoInfo.VideoTimestamp == null || VideoInfo.VideoTimestamp == new DateTime() ? null : VideoInfo.VideoTimestamp.ToString( "yyyy-MM-dd_HH-mm-ss" );
			string prefix = String.IsNullOrWhiteSpace( username ) ? datetime : String.IsNullOrWhiteSpace( datetime ) ? username : username + "_" + datetime;
			string basename = String.IsNullOrWhiteSpace( prefix ) ? title : ( prefix + "_" + title );

			return StringToFilename( basename, extension );
		}

		public override string GenerateOutputFilename() {
			return GetTargetFilename();
		}

		public override async Task<ResultType> Run( CancellationToken cancellationToken ) {
			JobStatus = VideoJobStatus.Running;
			Status = "Downloading...";

			string tempFoldername = Path.Combine(Util.TempFolderPath, GetTempFoldername());
			string urlFilename = "url.txt";
			string urlFilepath = Path.Combine(tempFoldername, urlFilename);
			string downloadFilename = "wip.bin";
			string downloadFilepath = Path.Combine(tempFoldername, downloadFilename);
			string progressFilename = "progress.txt";
			string progressFilepath = Path.Combine(tempFoldername, progressFilename);
			string movedFilename = "done.bin";
			string movedFilepath = Path.Combine(tempFoldername, movedFilename);
			string targetFilename = GetTargetFilename();
			string targetFilepath = Path.Combine(Util.TargetFolderPath, targetFilename);

			if (!await Util.FileExists(targetFilepath)) {
				if (!await Util.FileExists(movedFilepath)) {
					if (Directory.Exists(tempFoldername)) {
						Directory.Delete(tempFoldername, true);
					}
					Directory.CreateDirectory(tempFoldername);

					if (cancellationToken.IsCancellationRequested) { return ResultType.Cancelled; }

					File.WriteAllText(urlFilepath, VideoInfo.VideoId);

					var result = await ExternalProgramExecution.RunProgram(
						"wget",
						new string[] {
							"-i", urlFilename, "-o", progressFilename, "-O", downloadFilename
						},
						workingDir: tempFoldername,
						stdoutCallbacks: new System.Diagnostics.DataReceivedEventHandler[1] {
							(sender, received) => {
								if (!String.IsNullOrEmpty(received.Data)) {
									Status = received.Data;
								}
							}
						}
					);

					if (cancellationToken.IsCancellationRequested) { return ResultType.Cancelled; }
					File.Move(downloadFilepath, movedFilepath);
				}

				if (cancellationToken.IsCancellationRequested) { return ResultType.Cancelled; }
				await StallWrite(targetFilepath, new FileInfo(movedFilepath).Length, cancellationToken);
				if (cancellationToken.IsCancellationRequested) { return ResultType.Cancelled; }
				File.Move(movedFilepath, targetFilepath);
				Directory.Delete(tempFoldername, true);
			}

			Status = "Done!";
			JobStatus = VideoJobStatus.Finished;
			return ResultType.Success;
		}
	}
}

File: native/vodarchiver/videojobs/generic-file-job.cpp
#include "generic-file-job.h"

#include <atomic>
#include <cstddef>
#include <filesystem>
#include <format>
#include <string>
#include <string_view>

#include "util/file.h"
#include "util/text.h"

#include "vodarchiver/curl_util.h"
#include "vodarchiver/filename_util.h"

namespace VodArchiver {
static std::atomic<size_t> s_Counter = 0;

bool GenericFileJob::IsWaitingForUserInput() const {
    return false;
}

static std::string StringToFilename(std::string_view basename, std::string_view extension) {
    return std::format("{}.{}", MakeStringFileSystemSafeBaseName(basename), extension);
}

static std::string GetTempFoldername() {
    size_t c = s_Counter.fetch_add(1);
    return std::format("rawurl_{}", c);
}

static std::string GetTargetFilename(IVideoInfo& videoInfo) {
    // best guess...
    std::array<char, 256> buffer1;
    std::array<char, 256> buffer2;
    std::array<char, 256> buffer3;
    std::string_view videoId = videoInfo.GetVideoId(buffer1);
    std::string_view rawUsername = videoInfo.GetUsername(buffer2);
    std::string_view rawTitle = videoInfo.GetVideoTitle(buffer3);
    DateTime rawTimestamp = videoInfo.GetVideoTimestamp();
    std::string extension;
    {
        auto slash = HyoutaUtils::TextUtils::Split(HyoutaUtils::TextUtils::Trim(videoId), "/");
        if (!slash.empty()) {
            auto que = HyoutaUtils::TextUtils::Split(slash.back(), "?");
            if (!que.empty()) {
                auto underscore = HyoutaUtils::TextUtils::Split(
                    MakeStringFileSystemSafeBaseName(que.front()), "_");
                if (!underscore.empty()) {
                    extension = std::string(HyoutaUtils::TextUtils::Trim(underscore.back()));
                }
            }
        }
    }
    if (extension.empty()) {
        extension = "bin";
    }

    std::string_view username = HyoutaUtils::TextUtils::Trim(rawUsername);
    std::string_view fallbackTitle = "file";
    std::string_view title = HyoutaUtils::TextUtils::Trim(rawTitle).empty()
                                 ? fallbackTitle
                                 : HyoutaUtils::TextUtils::Trim(rawTitle);
    std::string datetime;
    if (!(rawTimestamp.Data == 0 || rawTimestamp.Data == DateTime::FromUnixTime(0).Data)) {
        datetime = DateTimeToStringForFilesystem(rawTimestamp);
    }
    std::string prefix = HyoutaUtils::TextUtils::Trim(username).empty() ? datetime
                         : HyoutaUtils::TextUtils::Trim(datetime).empty()
                             ? std::string(username)
                             : std::format("{}_{}", username, datetime);
    std::string basename = HyoutaUtils::TextUtils::Trim(prefix).empty()
                               ? std::string(title)
                               : std::format("{}_{}", prefix, title);

    return StringToFilename(basename, extension);
}

static std::string PathCombine(std::string_view lhs, std::string_view rhs) {
    std::string result(lhs);
    HyoutaUtils::IO::AppendPathElement(result, rhs);
    return result;
}

static bool DeleteDirectoryRecursive(const std::filesystem::path& rootDir) {
    {
        std::error_code ec;
        std::filesystem::directory_iterator iterator(rootDir, ec);
        if (ec) {
            return false;
        }
        while (iterator != std::filesystem::directory_iterator()) {
            if (iterator->is_directory()) {
                if (!DeleteDirectoryRecursive(*iterator)) {
                    return false;
                }
            } else {
                if (!HyoutaUtils::IO::DeleteFile(*iterator)) {
                    return false;
                }
            }
            iterator.increment(ec);
            if (ec) {
                return false;
            }
        }
    }

    return HyoutaUtils::IO::DeleteDirectory(rootDir);
}

static bool DeleteDirectoryRecursive(std::string_view path) {
    return DeleteDirectoryRecursive(HyoutaUtils::IO::FilesystemPathFromUtf8(path));
}

static ResultType RunGenericFileJob(GenericFileJob& job,
                                    JobConfig& jobConfig,
                                    TaskCancellation& cancellationToken) {
    std::unique_ptr<IVideoInfo> videoInfo;
    {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.JobStatus = VideoJobStatus::Running;
        job.TextStatus = "Downloading...";
        videoInfo = job.VideoInfo->Clone();
    }

    std::string tempFolderPath;
    std::string targetFolderPath;
    {
        std::lock_guard lock(jobConfig.Mutex);
        tempFolderPath = jobConfig.TempFolderPath;
        targetFolderPath = jobConfig.TargetFolderPath;
    }

    std::string tempFoldername = PathCombine(tempFolderPath, GetTempFoldername());
    std::string urlFilename = "url.txt";
    std::string urlFilepath = PathCombine(tempFoldername, urlFilename);
    std::string downloadFilename = "wip.bin";
    std::string downloadFilepath = PathCombine(tempFoldername, downloadFilename);
    std::string progressFilename = "progress.txt";
    std::string progressFilepath = PathCombine(tempFoldername, progressFilename);
    std::string movedFilename = "done.bin";
    std::string movedFilepath = PathCombine(tempFoldername, movedFilename);
    std::string targetFilename = GetTargetFilename(*videoInfo);
    std::string targetFilepath = PathCombine(targetFolderPath, targetFilename);

    if (!HyoutaUtils::IO::Exists(std::string_view(targetFilepath))) {
        if (!HyoutaUtils::IO::Exists(std::string_view(movedFilepath))) {
            if (HyoutaUtils::IO::DirectoryExists(std::string_view(tempFoldername))) {
                if (!DeleteDirectoryRecursive(std::string_view(tempFoldername))) {
                    return ResultType::Failure;
                }
            }
            if (!HyoutaUtils::IO::CreateDirectory(std::string_view(tempFoldername))) {
                return ResultType::Failure;
            }

            if (cancellationToken.IsCancellationRequested()) {
                return ResultType::Cancelled;
            }

            std::array<char, 256> buffer;
            std::string videoId(videoInfo->GetVideoId(buffer));
            if (!HyoutaUtils::IO::WriteFileAtomic(urlFilepath, videoId.data(), videoId.size())) {
                return ResultType::Failure;
            }

            auto response = VodArchiver::curl::GetFromUrlToMemory(videoId);
            if (!response || response->ResponseCode >= 400) {
                return ResultType::Failure;
            }
            if (!HyoutaUtils::IO::WriteFileAtomic(
                    downloadFilepath, response->Data.data(), response->Data.size())) {
                return ResultType::Failure;
            }

            if (cancellationToken.IsCancellationRequested()) {
                return ResultType::Cancelled;
            }

            if (!HyoutaUtils::IO::Move(downloadFilepath, movedFilepath)) {
                return ResultType::Failure;
            }
        }

        if (cancellationToken.IsCancellationRequested()) {
            return ResultType::Cancelled;
        }
        StallWrite(jobConfig,
                   targetFilepath,
                   HyoutaUtils::IO::GetFilesize(std::string_view(movedFilepath)).value_or(0),
                   cancellationToken,
                   ShouldStallWriteRegularFile,
                   [&](std::string status) {
                       std::lock_guard lock(*jobConfig.JobsLock);
                       job.TextStatus = std::move(status);
                   });
        if (cancellationToken.IsCancellationRequested()) {
            return ResultType::Cancelled;
        }
        if (!HyoutaUtils::IO::Move(movedFilepath, targetFilepath)) {
            return ResultType::Failure;
        }
        if (!DeleteDirectoryRecursive(std::string_view(tempFoldername))) {
            return ResultType::Failure;
        }
    }

    {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.TextStatus = "Done!";
        job.JobStatus = VideoJobStatus::Finished;
    }
    return ResultType::Success;
}

ResultType GenericFileJob::Run(JobConfig& jobConfig, TaskCancellation& cancellationToken) {
    return RunGenericFileJob(*this, jobConfig, cancellationToken);
}

std::string GenericFileJob::GenerateOutputFilename() {
    return GetTargetFilename(*this->VideoInfo);
}
} // namespace VodArchiver
